test/parallel/test-domain-promise.js

```diff
@@ -0,0 +1,25 @@
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const domain = require('domain');
+const vm = require('vm');
+
+const d = domain.create();
+
```

Comment:
@addaleax @trevnorris Here is the work I have so far on integrating promise hooks with async hooks: https://github.com/matthewloring/node/commit/dd178336a60c2698619b96a62267bd528c09ae0d#diff-5e552c79e1538215f1621d1774852e71R260. It is still very much a work in progress but the high level structure is all there. It provides both async hooks and domain propagation using the same mechanism and behavior as async wrap.